client/packages/programs/src/JsonForms/stripEmptyAdditions.ts

```diff
@@ -0,0 +1,109 @@
+import _ from 'lodash';
+import { JsonData } from './common';
+
+/**
+ * Checks if the input object or array is empty.
+ * An object is considered empty if all values are either undefined, empty objects, or arrays.
+ * For example, `{ value: undefined, obj: {} }`.
+ * An empty array this an array that is empty or only contains empty objects,
+ * e.g. `[{ value: undefined}]`.
+ */
+const objectOrArrayIsEmpty = (obj: JsonData | undefined): boolean => {
+  if (obj === undefined || obj === null) return true;
+  if (typeof obj !== 'object') return false;
+
+  // array
+  if (Array.isArray(obj)) {
+    if (obj.length === 0) return true;
+    return obj.every(it => objectOrArrayIsEmpty(it));
+  }
+
+  // object
+  if (Object.keys(obj).length === 0) return true;
+  const allValuesEmpty = Object.values(obj).every(it => {
+    if (typeof it === 'object') return objectOrArrayIsEmpty(it);
+    return it === undefined;
+  });
+  if (allValuesEmpty) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Recursively removes all empty data which has been added to the newData compared to the old
+ * input object.
+ *
+ * For example, given:
+ *
+ * old: { some: "value" }
+ * newData: { some: "value", obj: { add1: undefined }, array: [], add2: undefined}
+ *
+ * stripEmptyAdditions(old, newData) will return { some: "value" }.
+ */
```

Comment:
Just wondering -- why do you need to compare against `old` values? Why not just strip the empty/undefined values from the `newData` regardless of what `old` is?

Reply:
Because the old value could have empty objects, for example, from the import script.

I wanted to keep these empty unmodified object when saving the document.
Moreover, I only need to strip the new data. Otherwise I would need to strip old and new data and then compare them...